library/src/main/scala/scala/scalajs/reflect/Reflect.scala

```diff
@@ -0,0 +1,95 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js API               **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2017, LAMP/EPFL   **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+package scala.scalajs.reflect
+
+import scala.collection.mutable
+
+import scala.scalajs.js
+
+final class LoadableModuleClass[T] private[reflect] (
+    val runtimeClass: Class[T],
+    loadModuleFun: js.Function0[T]
+) {
+  /** Loads the module instance and returns it. */
+  def loadModule(): T = loadModuleFun()
+}
+
+final class InstantiatableClass[T] private[reflect] (
+    val runtimeClass: Class[T],
+    val declaredConstructors: List[InvokableConstructor[T]]
+) {
+  /** Instantiates a new instance of this class using the zero-argument
+   *  constructor.
+   *
+   *  @throws java.lang.InstantiationException (caused by a
+   *    `NoSuchMethodException`)
+   *    If this class is abstract or does not have a public zero-argument
+   *    constructor.
+   */
+  def newInstance(): T = {
+    declaredConstructors.find(_.parameterTypes.isEmpty).fold[T] {
+      throw new InstantiationException(runtimeClass.getName).initCause(
+          new NoSuchMethodException(runtimeClass.getName + ".<init>()"))
+    } { ctor =>
+      ctor.newInstance()
+    }
+  }
+}
+
+final class InvokableConstructor[T] private[reflect]  (
+    val parameterTypes: List[Class[_]],
+    newInstanceFun: js.Function
+) {
+  def newInstance(args: Any*): T = {
+    /* Check the number of actual arguments. We let the casts and unbox
+     * operations inside `newInstanceFun` take care of the rest.
+     */
+    require(args.size == parameterTypes.size)
+    newInstanceFun.asInstanceOf[js.Dynamic].apply(
+        args.asInstanceOf[Seq[js.Any]]: _*).asInstanceOf[T]
+  }
+}
+
+object Reflect {
+  /* I would like those val's to be `js.Dictionary`'es instead of full-blown
+   * Scala Maps. But if I do that, GCC "dead-code"-eliminates away statements
+   * that fill `loadableModuleClasses`! We fool it into not making a fool of
+   * itself by using more complicated data structures that it does not
+   * understand -_-'.
+   */
```

Comment:
*This*, is the ugliest thing I have ever done in this codebase. But honestly I don't know what else to do :(